architecture/fxa.md

```diff
@@ -0,0 +1,252 @@
+
+The userinfo endpoint validates the access token, and (upon success) provides at least the following information as "application/json":
+
+* `uid` - The user identifier string
+* `email` - The user's primary email address
+
+**WARNING**: The user's email address should not be stored unencrypted without their consent.
+
+## Browser Extension ##
+
+The Lockbox extension uses the async method `browser.identity.launchWebAuthFlow()` for OAuth-based sign in.  This method returns a Promise, providing the redirect URL when fulfilled.
+
+### Account States ###
+
+* `unbound` -- This Lockbox instance is not bound to FxA.  All data is stored locally using weaker protections; the encryption key is a default or platform-specific value and the datastore is accessible.
+* `unauthenticated` -- This Lockbox instance is bound to FxA, but the user is not signed into Lockbox; the encryption key is not known and the datastore is not accessible.
+* `authenticated` -- This Lockbox instance is bound to FxA and the user is signed into Lockbox; the encryption key is known and the datastore is accessible.
+
+### Initial State ###
+
+The a user first installs Lockbox, the datastore is initialized using a default master key.  If access to secure device storage is available, a random master key is generated and stored there.
+
+This allows a user to make use of Lockbox, retrieving/adding/updating/deleting entries.  However, none of this data is sent to remote cloud storage.
+
+### Binding ###
+
+When binding to an FxA account, the user clicks the "sign in" action (either from the full editor view or the condensed doorhanger view), which then triggers the following:
+
+1. The extension calls the `browser.identity.launchWebAuthFlow()` with the following members in the `details`:
+   - `interactive` is set to `true`
+   - `url` is set to FxA's authorization endpoint, including the query parameters for this specific request
+2. The browser opens a popup window requesting the user's FxA credentials
+3. Once the user submits their FxA credentials, FxA verifies the credentials and returns the authorization redirect URI containing the response's query parameters.
+4. The extension processes the authorization response and requests the initial token exchange
+5. FxA verifies the extension's request and (upon success) returns the token information
+6. The extension requests the user's profile information from the FxA userinfo endpoint, in order to obtain their FxA user identifier (`uid`)
+   - The access token is provided as authorization
+7. The FxA userinfo endpoint verifies the request and (upon success) returns the user's profile information
+8. The extension records the user identifier, access token, expiration time, and id token in `browser.storage.local` (the refresh token and application scoped key are only retained in memory)
```

Comment:
It may also be worth persisting the kid from the scoped key somewhere, so that you can easily detect if the keys have changed due to a password reset.

Reply:
That is a good point, and will get added in the next iteration of this doc.